### PR TITLE
Alerting: Extend PUT rule-group route to write the entire rule group rather than top-level fields only

### PR DESCRIPTION
**Pull request by alexweav**
**What this PR does / why we need it**:

Right now the PUT route for rule groups only allows updating the group-level fields (`for`, etc). This PR extends it to write the entire group resource. This is highly desirable for terraform and other clients, and also makes the route more RESTful. The existing GET route on the same endpoint already behaves this way.

This not being available breaks our terraform wrapper, which expects the resource to behave in a RESTful way.

**Which issue(s) this PR fixes**:

Rel https://github.com/grafana/grafana/issues/36153

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/api/api_provisioning.go [hunk 1]
```diff
@@ -56,7 +56,7 @@ type AlertRuleService interface {
 	UpdateAlertRule(ctx context.Context, rule alerting_models.AlertRule, provenance alerting_models.Provenance) (alerting_models.AlertRule, error)
 	DeleteAlertRule(ctx context.Context, orgID int64, ruleUID string, provenance alerting_models.Provenance) error
 	GetRuleGroup(ctx context.Context, orgID int64, folder, group string) (definitions.AlertRuleGroup, error)
-	UpdateRuleGroup(ctx context.Context, orgID int64, folderUID, rulegroup string, interval int64) error
+	ReplaceRuleGroup(ctx context.Context, orgID int64, group definitions.AlertRuleGroup, userID int64, provenance alerting_models.Provenance) error
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/api/api_provisioning.go:59]: I like the name a lot! It makes it more clear that it's a replacement and not some kind of update doing (perhaps) some magic.

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 2]
```diff
@@ -212,10 +308,61 @@ func (service *AlertRuleService) DeleteAlertRule(ctx context.Context, orgID int6
...
+		group.Rules[i].For = (time.Duration(group.Interval) * time.Second)
+		group.Rules[i].IntervalSeconds = group.Interval
+		group.Rules[i].RuleGroup = group.Title
+		group.Rules[i].NamespaceUID = group.FolderUID
+		group.Rules[i].OrgID = orgID
+	}
+	return group
+}
+
+func dropNilAlertRules(ptrs []*models.AlertRule) []models.AlertRule {
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/provisioning/alert_rules.go:362]: Is this not more of a transformation than a drop? I was confused by the naming at first.

**Review by JohnnyQQQQ — CHANGES_REQUESTED** [review]
I got one comment regarding `for`, overall looking good.

File: pkg/services/ngalert/provisioning/alert_rules.go [hunk 3]
```diff
@@ -212,10 +308,61 @@ func (service *AlertRuleService) DeleteAlertRule(ctx context.Context, orgID int6
...
+			service.log.Warn("failed to delete provenance record for rule: %w", err)
+		}
+	}
+	return nil
+}
+
+// syncRuleGroupFields synchronizes calculated fields across multiple rules in a group.
+func syncGroupRuleFields(group *definitions.AlertRuleGroup, orgID int64) *definitions.AlertRuleGroup {
+	for i := range group.Rules {
+		group.Rules[i].For = (time.Duration(group.Interval) * time.Second)
```
[the remark this answers is not in the document]
Comment (alexweav) [reply pkg/services/ngalert/provisioning/alert_rules.go:353]: Ah, good catch. Removed this and the corresponding test.

File: pkg/services/ngalert/provisioning/alert_rules.go (hunk 2 above)
Comment (alexweav) [reply pkg/services/ngalert/provisioning/alert_rules.go:362]: Renamed to `withoutNilAlertRules` so it's a little more fluent when calling as a transformation.

Comment (JohnnyQQQQ) [reply pkg/services/ngalert/provisioning/alert_rules.go]: I'm not sure we are talking about the same thing here. `append` will also append nil values? So it does not remove any nil alert rules?

Comment (JohnnyQQQQ) [reply pkg/services/ngalert/provisioning/alert_rules.go]: I would expect something like that in the beginning of the `for` loop 

```golang
if ptr == nil {
  continue
}
```

Comment (alexweav) [reply pkg/services/ngalert/provisioning/alert_rules.go]: Yep, totally forgot this :laughing: fixed!

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM 🚀

**Pull request merged**